src/components/collapsible-section-stateful/collapsible-section-stateful.tsx

```diff
@@ -0,0 +1,121 @@
+import { getState, setState } from '../../util/state-service';
+
+/**
+ * @slot - Content to put inside the collapsible section
+ * @exampleComponent limel-example-collapsible-section-stateful
+ */
+@Component({
+    tag: 'limel-collapsible-section-stateful',
+    shadow: true,
+})
+export class CollapsibleSectionStateful {
+    /**
+     * `true` if the section is expanded, `false` if collapsed.
+     */
+    @Prop({ mutable: true, reflect: true })
+    public isOpen: boolean = false;
+
+    /**
+     * Text to display in the header of the section
+     */
+    @Prop({ reflect: true })
+    public header: string;
+
+    /**
+     * Actions to place to the far right inside the header
+     */
+    @Prop()
+    public actions: Action[];
+
+    /**
+     * Used to identify the section for the purpose of remembering the
+     * open/closed state.
+     */
+    @Prop({ reflect: true })
+    public stateKey: string;
+
+    /**
+     * Emitted when the section is expanded
+     */
+    @Event()
+    private open: EventEmitter<void>;
+
+    /**
+     * Emitted when the section is collapsed
+     */
+    @Event()
+    private close: EventEmitter<void>;
+
+    /**
+     * Emitted when an action is clicked inside the header
+     */
+    @Event()
+    private action: EventEmitter<Action>;
+
+    private stateId: string;
+
+    public connectedCallback() {
+        if (this.stateKey) {
+            this.stateId = `limel-collapsible-section-stateful.${this.stateKey}.isOpen`;
+
+            const isOpen = getState(this.stateId);
+
+            if (
+                (isOpen === true || isOpen === false) &&
+                this.isOpen !== isOpen
+            ) {
+                this.isOpen = isOpen;
+                this.emitOpenState();
+            }
+        }
+    }
+
+    public render() {
+        return (
+            <limel-collapsible-section
+                isOpen={this.isOpen}
+                header={this.header}
+                actions={this.actions}
+                onOpen={this.handleOpenEvent}
+                onClose={this.handleCloseEvent}
+                onAction={this.handleActionEvent}
+            >
+                <slot />
+            </limel-collapsible-section>
+        );
+    }
+
+    @Watch('isOpen')
+    protected openCloseHandler(value: boolean) {
+        if (this.stateId) {
+            setState(this.stateId, value);
+        }
+    }
+
+    private emitOpenState = () => {
+        if (this.isOpen) {
+            this.open.emit();
+        } else {
+            this.close.emit();
+        }
+    };
+
+    private handleOpenEvent = () => {
+        if (!this.isOpen) {
+            this.isOpen = true;
+        }
+    };
+
+    private handleCloseEvent = () => {
+        if (this.isOpen) {
+            this.isOpen = false;
+        }
+    };
+
+    private handleActionEvent = (event: CustomEvent) => {
+        event.stopPropagation();
+        this.action.emit(event.detail);
```

Comment:
I hate this, but I haven't found any way to get the `onAction` property to show up on the element's type, other than declaring an EventEmitter on the component, and then the compiler complains about the property being unused, so we end up with this. Any ideas?

Reply:
Yeah it's a bit annoying that you have to re-declare any custom events from inner components. I get why because of the shadow dom encapsulation, but it's also weird since all events bubble.

I've usually solved this in the past by declaring the event as `protected` just to get the linter to not complain about unused properties :sweat_smile:

Reply:
Ah! Of course! I even introduced that trick myself for `@Watch()` handlers, but didn't think to use it here 🤦‍♂️ 😄

Thank you! 🙏

Reply:
⚡